cunumeric/deferred.py

```diff
@@ -305,37 +305,278 @@ def get_scalar_array(self):
+                + b_shape
+                + tuple(
+                    self.shape[i] for i in range(start_index + N, self.ndim)
+                )
+            )
+            new_arrays = tuple()
+            # promote all index arrays to have the same shape as output
+            for a in arrays:
+                for i in range(0, start_index):
+                    a = a.promote(i, self.shape[i])
+                for i in range(start_index + N, self.ndim):
+                    a = a.promote(key_dim + i - N, self.shape[i])
+                new_arrays += (a,)
+            arrays = new_arrays
+
+        else:
+            # the use case when # of arrays passed is equal to the dimension
+            # of the input array
+            if len(arrays) > self.ndim:
+                raise ValueError("wrong number of index arrays passed")
+            new_arrays = tuple()
+            for a in arrays:
+                if data_type != a.dtype:
+                    raise TypeError(
+                        "type of all index arrrays should be the same"
+                    )
+                if a.shape != b_shape:
+                    a = a._broadcast(b_shape)
+                else:
+                    a = a.base
+                new_arrays = new_arrays + (a,)
+        arrays = new_arrays
+
+        # create output array which will store Point<N> field where
+        # N is number of index arrays
+        # shape of the output array should be the same as the shape of each
+        # index array
+        # NOTE: We need to instantiate a RegionField of non-primitive
+        # dtype, to store N-dimensional index points, to be used as the
+        # indirection field in a copy.
+        # Such dtypes are technically not supported,
+        # but it should be safe to directly create a DeferredArray
+        # of that dtype, so long as we don't try to convert it to a
+        # NumPy array.
+        N = self.ndim
+        pointN_dtype = self.runtime.get_point_type(N)
+        store = self.context.create_store(
+            pointN_dtype, shape=out_shape, optimize_scalar=True
+        )
+        output_arr = DeferredArray(
+            self.runtime, base=store, dtype=pointN_dtype
+        )
+
+        # call ZIP function to combine index arrays into a singe array
+        task = self.context.create_task(CuNumericOpCode.ZIP)
+        task.add_output(output_arr.base)
+        if len(arrays) < self.ndim:
+            task.add_scalar_arg(self.ndim, ty.int64)  # N of points in Point<N>
+            task.add_scalar_arg(key_dim, ty.int64)  # key_dim
+            task.add_scalar_arg(start_index, ty.int64)  # start_index
+            for a in arrays:
+                task.add_input(a)
+                task.add_alignment(output_arr.base, a)
+                task.add_broadcast(a, axes=tuple(range(1, len(out_shape))))
+                task.add_broadcast(
+                    output_arr.base, axes=tuple(range(1, len(out_shape)))
+                )
```

Comment:
Why are you broadcasting here?

Reply:
I think I don't need to anymore. It is a leftover from my original implementation